[DBZ-706] VERY rough prototype code

Crim opened the pull request:
Some extremely rough prototype code tossed together to determine if this was possible to build.

Questions I have:
- Is the SourceInfo instance the right place to be storing the original Query?
- I feel like I'm missing places where the query property should be cleared out between RBR events.

Other comments:
- For simplicity in this case I've just enabled the binlog_rows_query_log_events option globally, but this also exists as a session variable, so tests could/should be added around just selectively toggling the option and validating that the events get parsed correctly.

gunnarmorling commented:
One other observation: I see `"query" : null` in emitted messages if no query could be determined. It'd be nice to just omit the `query` attribute in that case.

Crim commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SourceInfo.java:
```diff
@@ -127,6 +128,8 @@
                                                      .field(THREAD_KEY, Schema.OPTIONAL_INT64_SCHEMA)
                                                      .field(DB_NAME_KEY, Schema.OPTIONAL_STRING_SCHEMA)
                                                      .field(TABLE_NAME_KEY, Schema.OPTIONAL_STRING_SCHEMA)
+                                                     // TODO how to make this optional?
```
Since this is built statically, is it possible to alter the schema instance to drop the field when it has no value?  I thought we could modify it in the schema() method below, but that seems like building in a fairly nasty side effect based on currentQuery being null or not.

gunnarmorling replied:
The schema would always contain "query" as an optional String field. But the code for creating the struct would only populate that field if there's a non-null value.

Crim replied:
So this would seem to be doing that correctly then? https://github.com/debezium/debezium/pull/523/files#diff-3a15e1a239968077827d2be8142eb3cdR348

gunnarmorling replied:
Yes, that code looks alright. Has this changed since yesterday? Because I definitely saw `"query ": null` in messages when building the connector with that PR and running it in Kafka Connect.

Crim replied:
Same code as yesterday, and no you aren't crazy :) . The tests I added confirm that `"query": null` is present in the messages if that field is not set.

Crim replied:
Let me re-run the test cases w/ a breakpoint and just verify for myself that the statement is evaluating as false

gunnarmorling replied:
Ok, that's re-assuring that I didn't dream it :)

Crim replied:
Hmm maybe **I'm** the crazy one, running through the test cases that should not include the query property, it's not showing up in the SourceRecord.toString() output.  Perhaps I didn't have the if statement yesterday after all?

I'm about to head out for the day, but I can spend some additional time tomorrow and validate this is working as expected across more test cases.

Crim replied:
OK took another look at this.  When the query field is not populated the resulting struct looks like:

![image](https://user-images.githubusercontent.com/571653/40759272-5fe87ad6-64cc-11e8-8b22-190f722fb1ad.png)

Compared to when it IS populated with a value:

![image](https://user-images.githubusercontent.com/571653/40759307-7cbfcfa6-64cc-11e8-8d33-69b5b94c3e17.png)

I believe this matches with the behavior you were looking for correct?

gunnarmorling replied:
Yeah, all looks good now. I'm going to remove the TODO comment before merging.

Crim commented on debezium-core/src/test/java/io/debezium/data/VerifyRecord.java:
```diff
@@ -250,6 +250,77 @@ public static void isValidTombstone(SourceRecord record, String pkField, int pk)
         isValidTombstone(record);
     }
 
+    /**
```
I've shuffled some code around from AbstractConnectorTest, if this isn't the right approach here happy to back it out.

gunnarmorling replied:
Seems good to me :)

Crim commented:
Hey @gunnarmorling ,

I've made some updates to the PR per your suggestions.

- Reverted the debugging statements I added to get my bearings as well as the my.cnf configuration changes.
- Added a new configuration option to enable including the SQL query in the event.  I set the default to NOT include it, as well as added a warning to its description about the potential for leaking masked or filtered columns.
- Added test coverage.

Crim commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java:
```diff
@@ -265,6 +266,11 @@ protected void doStart() {
         eventHandlers.put(EventType.XA_PREPARE, this::prepareTransaction);
         eventHandlers.put(EventType.XID, this::handleTransactionCompletion);
 
+        // Conditionally register QUERY handler to parse SQL statements.
```
TODO: Comment should read '.... ROWS_QUERY handler ....'

Crim commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java:
```diff
@@ -573,6 +579,20 @@ protected void handleGtidEvent(Event event) {
     }
 
+    /**
+     * Handle the supplied event with an {@link RowsQueryEventData} by possibly recording the original SQL query
```
TODO: Comment should be reworded.

gunnarmorling commented on debezium-core/src/test/java/io/debezium/data/VerifyRecord.java:
```diff
@@ -250,6 +250,77 @@ public static void isValidTombstone(SourceRecord record, String pkField, int pk)
+            String fieldName = fieldNames[i];
+            if (value instanceof Struct) {
+                value = ((Struct)value).get(fieldName);
+            } else {
```
Minor nit on style: can you move the "else {" to the next line, i.e. there should be no code after closing braces. I know it's not done consistently in the code base, but with Jiri we recently agreed on using that style going forward in any new or changed code. Thanks!

Crim commented:
I went to update the documentation on https://github.com/debezium/debezium.github.io but ran into some trouble with the docker image in that project:

```bash
awestruct@eefc0396fa78:/site$ rake clean preview
sed: can't read vendor/bundle/ruby/2.3.0/gems/awestruct-0.5.7/lib/awestruct/page.rb: No such file or directory
/usr/local/bundle/bin/bundle:104:in `load': cannot load such file -- /usr/local/lib/ruby/gems/2.3.0/gems/bundler-1.16.2/exe/bundle (LoadError)
    from /usr/local/bundle/bin/bundle:104:in `<main>'
```

Crim commented:
Updated Documentation in this PR: https://github.com/debezium/debezium.github.io/pull/151

gunnarmorling commented on debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/MySqlConnectorIT.java:
```diff
@@ -1102,8 +1105,507 @@ public void shouldEmitNoTombstoneOnDelete() throws Exception {
+     */
+    @Test
+    @FixFor("DBZ-706")
+    public void parseMultipleRowInsertStatement() throws Exception {
```
Really great and thorough testing, thanks a lot!

Crim commented:
Perfect!  Thanks!
